pkg/credentialprovider/azure_credentials.go

```diff
@@ -67,29 +66,56 @@ type acrProvider struct {
 	registryMirror map[string]string // Registry mirror relation: source registry -> target registry
 }
 
-func NewAcrProvider(config *providerconfig.AzureClientConfig, environment *azclient.Environment, credential azcore.TokenCredential) CredentialProvider {
-	return &acrProvider{
-		config:      config,
-		credential:  credential,
-		environment: environment,
-	}
-}
+type getTokenCredentialFunc func(req *v1.CredentialProviderRequest, config *providerconfig.AzureClientConfig) (azcore.TokenCredential, error)
 
-// NewAcrProvider creates a new instance of the ACR provider.
-func NewAcrProviderFromConfig(configFile string, registryMirrorStr string) (CredentialProvider, error) {
-	if len(configFile) == 0 {
-		return nil, errors.New("no azure credential file is provided")
-	}
+func NewAcrProvider(req *v1.CredentialProviderRequest, registryMirrorStr string, configFile string) (CredentialProvider, error) {
 	config, err := configloader.Load[providerconfig.AzureClientConfig](context.Background(), nil, &configloader.FileLoaderConfig{FilePath: configFile})
 	if err != nil {
 		return nil, fmt.Errorf("failed to load config: %w", err)
 	}
 
+	_, env, err := azclient.GetAzCoreClientOption(&config.ARMClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	clientOption, _, err := azclient.GetAzCoreClientOption(&config.ARMClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	var getTokenCredential getTokenCredentialFunc
+
+	// kubelet is responsible for checking the service account token emptiness when service account token is enabled, and only when service account token provide is enabled,
+	// service account token is set in the request, so we can safely check the service account token emptiness to decide which credential to use.
+	if len(req.ServiceAccountToken) != 0 {
```

Comment:
What is SA is set without proper annotation `kubernetes.azure.com/acr-client-id`, should we fall back to MI automatically?

Reply:
once the serviceaccount token is set, we will treat the serviceaccount feature is enabled, and the user specify the required  annotation, to achieve a secure acr authentication. Falling back to MI is against the security intetion.